### PR TITLE
Vectorized, torch.compile-compatible Ewald summation, added docs

### PR DESCRIPTION
**Pull request by djkim-kr**
## What this adds

A **vectorized Ewald summation** that `torch.compile` and AOTInductor can trace, so LES models
can now be compiled for training and exported for deployment. The legacy implementation is
untouched and still the default.

### `Ewald_vectorized`

`src/les/module/ewald_vectorized.py` is a new class next to the existing `Ewald`, with the same
interface. Two things made the old one untraceable, and both are gone:

- **Python loop over the structures in a batch.** The new class evaluates the whole batch at
  once, in one code path.
- **Branching on `det(cell)` to decide periodicity per structure.** That is a decision taken
  from tensor *values*, which stops a trace. The boundary condition is now fixed at
  construction by `is_periodic`.

Kernels for the multipole terms live in `make_kernels_vectorized.py` so the monopole, dipole,
quadrupole and induced terms are built separately rather than inlined.

A few smaller choices follow from what AOTInductor can lower: scatter/gather is replaced by
one-hot `matmul` (its CPU backend cannot vectorize the `atomic_add` a scatter lowers to), and
masking uses `torch.where` rather than multiplication, since `0 * inf` is a sticky `NaN`.
`forward()` is TorchScript-friendly, and older checkpoints still deserialize.

### Choosing an implementation

```python
Les({"is_periodic": True})    # vectorized, periodic      -- compilable
Les({"is_periodic": False})   # vectorized, non-periodic  -- compilable
Les({})                       # legacy, decides per structure -- not compilable
```

`is_periodic` defaults to `None`, so **existing code keeps the legacy module and behaves
exactly as before**. `N_max` (default 10) bounds the k-grid; keep `N_max * dl` above the cell's
longest side.

Supported terms: monopoles, dipoles, quadrupoles, induced charges (`-iq`) and induced dipoles
(`-iu`), periodic and non-periodic.

## Tests

`test/run_all.py` runs everything and prints one row per target:

```bash
python test/run_all.py
```

New test files:

| file | what it checks |
|---|---|
| `test_ewald_vectorized_physics.py` | vectorized vs legacy, term by term -- monopole, dipole, quadrupole, induced charge, induced dipole, BEC, periodic and non-periodic |
| `test_ewald_vectorized_compile.py` | `torch.compile` and AOTInductor export gates, including gradients of the latent charges |
| `test_ewald_vectorization_compilation.py` | legacy vs vectorized agreement |
| `test_ewald_vectorized_all.py` | all terms together |
| `test_torch_all_features.py` | TorchScript, every term enabled |
| `_vec_harness.py` | shared harness the vectorized tests are built on |

The physics tests are the important ones: the vectorized implementation is checked **against the
legacy one**, so the existing behaviour is the reference. Compile gates that a given PyTorch
version cannot support are skipped with a reason rather than failing.

All targets pass on the most recent PyTorch 2.13.0 both on CPU and CUDA.

## Documentation

`docs/` is new, published at [les.readthedocs.io](https://les.readthedocs.io): how LES works, the
library API, and per-MLIP guides for MACE, CACE, NequIP and Allegro.

## Compatibility

- Default behaviour is unchanged: without `is_periodic`, the legacy `Ewald` is used.
- Older serialized models still load; the new attributes get their historical defaults.
- The `Les` forward signature and outputs are unchanged.

### CODE REVIEW

**Comment by BingqingCheng** [comment]
Thanks for this substantial update. The vectorized implementation looks carefully constructed, and the comparisons against the legacy Ewald implementation—covering energies, forces, multipoles, induced terms, non-unit `sigma`/`dl`, and both self-interaction settings—give me reasonable confidence in the intended physics equivalence.

I have two concerns that I think should be addressed before merging:

1. **The AOTInductor compatibility claim is broader than what the tests enforce.** Periodic AOT export with gradients of the latent multipoles is listed in `KNOWN_GAPS`, so failure of this path does not fail the suite. This seems inconsistent with the claims that the implementation is fully compile/AOT compatible and that all targets pass. Please either fix this path or explicitly narrow the documented guarantee to what is actually supported—for example, energy and `dE/dr` for deployment, but not AOT-exported gradients with respect to all latent multipoles. The test summary should also distinguish known gaps from passes.

2. **The test runner can report false-positive passes.** It treats script targets as passing whenever they exit with code zero, but some scripts only print comparisons or catch and log exceptions. For example, `test_ewald_vectorization_compilation.py` prints the legacy and vectorized results without asserting equality, while `test_torch.py` catches scripting failures without exiting nonzero. Please use assertions or `torch.testing.assert_close` and propagate exceptions, or convert these checks to pytest tests, so that a green suite means every advertised check actually succeeded.

A few additional points should also be addressed or clarified:

- Please add a regression test that loads an artifact created before the new attributes. This would substantiate the older-checkpoint compatibility claim and clarify whether it covers whole-module serialization, TorchScript, `state_dict`s, or all three.
- Please document prominently that `is_periodic=True/False` fixes one boundary condition for the entire `Les` instance and therefore cannot handle mixed-periodicity batches; only the legacy `None` path supports that case.
- The fixed `N_max` grid can silently omit reciprocal vectors. The criterion `N_max * dl > longest cell side` is a useful orthorhombic heuristic, but not a general guarantee for strongly skewed triclinic cells. Please qualify this guidance and recommend a convergence check; a test with a more strongly skewed cell would also be useful.
- `docs/library.md` says that `compute_field` returns the per-atom potential and field, but `Les.forward()` does not expose `phi` or `field` in its output dictionary. Please correct the documentation or expose those outputs.
- There is an incomplete sentence in `docs/theory.md`: “Because \(E^\mathrm{lr}\) depends on them only through a physical …”.
- The PR description says that scatter/gather was replaced by one-hot matrix multiplication, although some `scatter_add_` operations remain. Please describe more narrowly the particular problematic path that was replaced.

Overall, I think the physics implementation is promising. My request for changes is mainly about defining the compilation guarantee precisely and ensuring that the validation suite cannot mask failures.

**Comment by djkim-kr** [comment]
address review: real assertions, no known-gap escape hatch

- drop `KNOWN_GAPS`; the periodic latent-grad AOTI gate is no longer run,
  since deployment differentiates positions only (gates C/E cover that)
- assert the legacy-vs-vectorized comparisons in the `tests` instead of only printing them
- `run_all.py` distinguishes PASS (numbers checked) from RAN (smoke only)
- remove `test_torch.py`; superseded by `test_torch_all_features.py`
- `docs`: mixed-periodicity limit, N_max criterion, compute_field outputs,
  complete a truncated sentence in theory.md
